### ISSUE TITLE
Name constructor required 3 arguments prior to 0.3, and 4 arguments after

### ISSUE THREAD

**Issue by mdanatg** [serge-sans-paille/gast#37]
It looks like the new grammar added a 4th `type_comment` argument to `Name` nodes. Upgrading the code for this change would break existing users and force them to upgrade their gast installation. Is it possible to default this argument to `None` instead? Then the code would remain compatible across versions.

**Comment by serge-sans-paille** [comment]
Yeah, that's python3.8 ast that goes this way. And gast tries hard to follow upstream ast update, while  still being able to represent older version. That's for the change motivation :-)

Note that because `ast.Num`, `ast.Str` and the likes nodes disappeared in favor of `ast.Constant`, a code base change is still needed, as advertised by the version bump. Same goes for `ast.For`, `ast.Module` etc.

So maybe it's a good thing that code breaks? py38 actually brings in quite a lot of ast changes? It's still possible to freeze version to '0.2.2' if you don't care about py38.

I'd like to here more of your thoughts about this. I understand the fix you propose, but I'm wondering if it's the right approach... On one hand, providing None / empty list defaults makes sense for several nodes, on the other hand this introduces some kind of difference with the ast module API... Maybe we should provide a `gast.gastx` module with enhanced API?

**Comment by mdanatg** [comment]
This makes sense. For now, we pinned the version to 0.2.2 in TF and I think we can later package together the necessary code updates with advancing the version to 0.3.2.

In general, I agree that breaking changes are preferable if they allow for a cleaner design, and that offsets the convenience of automatically using the latest version.

One thing worth noting is that `ast` does seem to have these defaults in its API: for example, `ast.Name('a', as.Load())` works and fills the other attribute with defaults. But I'm not sure that matching the standard API, in addition to the grammar, is worth the extra complexity.

**Comment by serge-sans-paille** [comment]
> One thing worth noting is that ast does seem to have these defaults in its API:
> for example, ast.Name('a', as.Load()) works and fills the other attribute with
> defaults. But I'm not sure that matching the standard API, in addition to the
> grammar, is worth the extra complexity.

Actually, in py38, the ast.Name object only has two positional parameters.
That's a point where gast and ast differ, it's docummented in the README, and
there's one reason behind this: py2 uses expression type to represent individual
argument, so that

    def foo((x, y)): pass

is valid. It's no longer possible in py3x, and as a result, argument are
represented using arg node. The only common representation I could find was to
extend the ast.Name node to have extra attributes :-/

**Comment by mdanatg** [comment]
I think the current state is ok, it's not bothersome once you switch to 0.3+
